File: resources/recipes/data/parse-log-files/parse_log.js
const fs = require('fs');
const readline = require('readline');

const LOG_PATTERN = /^(\S+) \S+ \S+ \[([^\]]+)\] "(\S+) (\S+) ([^"]+)" (\d{3}) (\S+)/;

async function parseApacheLog(filePath) {
    const stream = fs.createReadStream(filePath);
    const rl = readline.createInterface({ input: stream });
    const statusCounts = {};
    let total = 0;
    let parseErrors = 0;

    for await (const line of rl) {
        total += 1;
        const match = LOG_PATTERN.exec(line);
        if (match) {
            const status = match[6];
            statusCounts[status] = (statusCounts[status] || 0) + 1;
        } else {
            parseErrors += 1;
            console.log(`malformed line ${total}: ${line.slice(0, 120)}`);
        }
    }

    console.log('\nStatus counts:', statusCounts);
    console.log(`Parse errors: ${parseErrors}/${total} (${(100 * parseErrors / total).toFixed(2)}%)`);
}

parseApacheLog('access.log').catch(console.error);

File: resources/recipes/data/parse-log-files/parse_log.py
import re
from collections import Counter

LOG_PATTERN = re.compile(
    r'(?P<ip>\S+) \S+ \S+ \[(?P<time>[^\]]+)\] '
    r'"(?P<method>\S+) (?P<path>\S+) (?P<proto>[^"]+)" '
    r'(?P<status>\d{3}) (?P<bytes>\S+)'
)


def parse_apache_log(path):
    status_counts = Counter()
    parse_errors = 0
    total = 0

    with open(path, 'r', encoding='utf-8') as f:
        for line in f:
            total += 1
            match = LOG_PATTERN.match(line)
            if match:
                status_counts[match.group('status')] += 1
            else:
                parse_errors += 1
                print(f'malformed line {total}: {line.rstrip()[:120]}')

    print('\nStatus counts:', dict(status_counts))
    print(f'Parse errors: {parse_errors}/{total} ({100 * parse_errors / total:.2f}%)')


if __name__ == '__main__':
    parse_apache_log('access.log')
